pkg/credentialprovider/aws/aws_credentials.go

```diff
@@ -17,25 +17,65 @@ limitations under the License.
 package credentials
 
 import (
+	"bytes"
 	"encoding/base64"
+	"errors"
 	"fmt"
+	"io/ioutil"
+	"net/http"
 	"strings"
 	"time"
 
 	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/ec2metadata"
+	"github.com/aws/aws-sdk-go/aws/endpoints"
 	"github.com/aws/aws-sdk-go/aws/request"
 	"github.com/aws/aws-sdk-go/aws/session"
 	"github.com/aws/aws-sdk-go/service/ecr"
 	"k8s.io/klog"
 
+	"k8s.io/apimachinery/pkg/util/sets"
 	"k8s.io/kubernetes/pkg/credentialprovider"
 	"k8s.io/kubernetes/pkg/version"
 )
 
-const awsChinaRegionPrefix = "cn-"
-const awsStandardDNSSuffix = "amazonaws.com"
-const awsChinaDNSSuffix = "amazonaws.com.cn"
-const registryURLTemplate = "*.dkr.ecr.%s.%s"
+const (
+	awsChinaRegionPrefix        = "cn-"
+	awsStandardDNSSuffix        = "amazonaws.com"
+	awsChinaDNSSuffix           = "amazonaws.com.cn"
+	registryURLTemplate         = "*.dkr.ecr.%s.%s"
+	awsOsakaLocalRegion         = "ap-northeast-3"
+	awsAvailabilityZoneEndpoint = "placement/availability-zone"
+	awsChassisFile              = "/sys/devices/virtual/dmi/id/chassis_vendor"
+	awsHypervisorFile           = "/sys/hypervisor/uuid"
+	awsVersionFile              = "/sys/hypervisor/version/extra"
+)
+
+func init() {
+	seen := sets.NewString()
+
+	for _, p := range endpoints.DefaultPartitions() {
+		for r := range p.Regions() {
+			if !seen.Has(r) {
+				registerCredentialsProvider(r)
+				seen.Insert(r)
+			}
+		}
+	}
+
+	// ap-northeast-3 is purposely excluded from the SDK because it requires an
+	// access request see: https://github.com/aws/aws-sdk-go/issues/1863
+	registerCredentialsProvider(awsOsakaLocalRegion)
+
+	// Register the current region if we haven't seen it already just in case
+	// we're in a region that's not in the SDK for some reason.
+	currentRegion, err := getCurrentRegion()
+	if err != nil {
+		klog.Warningf("unable to detect region for current host, error: %v", err)
```

Comment:
as an example why doing things like this in init blocks is less than ideal, building this PR and running `hyperkube kubectl` produces this:

```
$ hyperkube kubectl
W0208 21:08:15.924885   92826 aws_credentials.go:74] unable to detect region for current host, error: RequestError: send request failed
caused by: Get http://169.254.169.254/latest/meta-data/placement/availability-zone: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers)
kubectl controls the Kubernetes cluster manager. 

Find more information at: https://kubernetes.io/docs/reference/kubectl/overview/

Basic Commands (Beginner):
...
```

Reply:
Agreed this is not ideal... we wouldn't run into this error case actually if we assumed the SDK returned all valid regions but it seems this check was added as a safety measure. @mcrute any ideas?

Reply:
I agree we should stay away from init blocks, but that would require a bit of refactoring on credential providers and this seems in line with the way the other providers do it. I think we should at least make sure the init block isn't going to spew a bunch of errors on non-AWS instances though.

Reply:
Passive registration on init seems ok, with no network/file checks and no errors. The active checks would move into Enabled().

Reply:
Sounds good to me! So for this case we would have to run the current region check in `Enabled()` (or somewhere else?) or remove the current region check altogether assuming the SDK returns all valid regions.

Reply:
I updated the code to only check the metadata if running on an AWS instance. This exists to catch a pretty rare (though possible) edge-case. I don't think it can be removed completely.

Reply:
I think the concern here is that we'll still do the file check as part of `onAwsInstance` (which logs on err) as part of an init block which adds noise and unnecessary calls to components that import the credential provider but doesn't need to use it (hyperkube is an example as mentioned by @liggitt above https://github.com/kubernetes/kubernetes/pull/70675#discussion_r255283077). 

I guess in this scenario, the worst-case scenario for non-AWS instances is 1 file check. Not ideal but maybe worth it, thoughts @liggitt? 

Wouldn't be a problem if the credential provider was out-of-tree but that's another problem :P

Reply:
On the logging front I could either remove the log statements completely from `onAWSInstance` or push them to a higher log level. I think pushing to a higher level is the best bet but open to other ideas.